Web/Scotty/Action.hs

```diff
@@ -354,6 +364,25 @@ jsonData = do
         A.Error err -> throwIO $ FailedToParseJSON b $ T.pack err
         A.Success a -> return a
 
+-- | Parse the request body as @x-www-form-urlencoded@ form data and return it.
+--
+--   The form is parsed using 'urlDecodeAsForm'. If that returns 'Left', the
+--   status is set to 400 and an exception is thrown.
+formData :: (FromForm a, MonadUnliftIO m) => ActionT m a
+formData = do
+  form <- paramListToForm <$> formParams
+  case fromForm form of
+    Left err -> throwIO $ MalformedForm err
+    Right value -> return value
+  where
+    -- This rather contrived implementation uses cons and reverse to avoid quadratic complexity (e.g. using HashMap.insertWith (++)).
+    -- It iterates over all parameters, prepending values for duplicate keys and reverses all hashmap entries afterwards.
+    paramListToForm :: [Param] -> Form
+    paramListToForm = Form . fmap reverse . foldl' (\f (k, v) -> HashMap.alter (prependValue v) k f) HashMap.empty
```

Comment:
> It iterates over all parameters, prepending values for duplicate keys and reverses all hashmap entries afterwards.

this describes the implementation, but a comment should explain _why_ we need this construction in the first place.

Reply:
Ah, the reason is given in the first line:

> to avoid quadratic complexity (e.g. using HashMap.insertWith (++))

Reply:
Perhaps I could change it to "avoid quadratic complexity [...] when constructing a `Form` from a list of `Param`s.

Reply:
Yes this would be helpful!